protocols/bftcosi/bftcosi.go

```diff
@@ -0,0 +1,615 @@
+		bft.tpcMut.Lock()
+		bft.tempPrepareCommit = append(bft.tempPrepareCommit, comm.Commitment)
+		if len(bft.tempPrepareCommit) < len(bft.Children()) {
+			bft.tpcMut.Unlock()
+			return nil
+		}
+		commitment = bft.prepare.Commit(bft.tempPrepareCommit)
+		bft.tpcMut.Unlock()
+		if bft.IsRoot() {
+			return bft.startChallengePrepare()
+		}
+
+		dbg.Lvl3(bft.Name(), "BFTCoSi handle Commit PREPARE")
+	case RoundCommit:
+		bft.tccMut.Lock()
+		bft.tempCommitCommit = append(bft.tempCommitCommit, comm.Commitment)
+		if len(bft.tempCommitCommit) < len(bft.Children()) {
+			bft.tccMut.Unlock()
+			return nil
+		}
+		commitment = bft.commit.Commit(bft.tempCommitCommit)
+		bft.tccMut.Unlock()
+		if bft.IsRoot() {
+			// do nothing:
+			// stop the processing of the round, wait the end of
+			// the "prepare" round: calls startChallengeCommit
+			return nil
+		}
+
+		dbg.Lvl3(bft.Name(), "BFTCoSi handle Commit COMMIT")
+	}
+	// set same RoundType as for the received commitment:
+	typedCommitment.TYPE = comm.TYPE
+	typedCommitment.Commitment = commitment
+	err := bft.SendTo(bft.Parent(), typedCommitment)
+	return err
+}
+
+// startPrepareChallenge create the challenge and send its down the tree
+func (bft *ProtocolBFTCoSi) startChallengePrepare() error {
+	// create challenge from message's hash:
+	prep := &msgPrepare{Msg: bft.Msg}
+	h := prep.Hash()
+	ch, err := bft.prepare.CreateChallenge(h)
+	if err != nil {
+		return err
+	}
+
+	bftChal := &ChallengePrepare{
+		TYPE:      RoundPrepare,
+		Challenge: ch,
+		Msg:       bft.Msg,
+	}
+
+	go bft.verificationFun(bft.Msg, bft.verifyChan)
+
+	dbg.Lvl3(bft.Name(), "BFTCoSi Start Challenge PREPARE")
+	// send to children
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, bftChal)
+	}
+	return err
+}
+
+// startCommitChallenge waits the end of the "prepare" round.
+// Then it creates the challenge and sends it along with the
+// "prepare" signature down the tree.
+func (bft *ProtocolBFTCoSi) startChallengeCommit() error {
+	c, err := bft.commit.CreateChallenge(bft.Msg)
+	if err != nil {
+		return err
+	}
+
+	// send challenge + signature
+	cc := &ChallengeCommit{
+		TYPE:      RoundCommit,
+		Challenge: c,
+		Signature: bft.prepare.Signature(),
+	}
+	dbg.Lvl3("BFTCoSi Start Challenge COMMIT")
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, cc)
+	}
+	return err
+}
+
+// handlePrepareChallenge receive the challenge messages for the "prepare"
+// round.
+func (bft *ProtocolBFTCoSi) handleChallengePrepare(ch *ChallengePrepare) error {
+	bft.Msg = ch.Msg
+	// start the verification of the message
+	go bft.verificationFun(bft.Msg, bft.verifyChan)
+	// acknowledge the challenge and send its down
+	chal := bft.prepare.Challenge(ch.Challenge)
+	ch.Challenge = chal
+
+	dbg.Lvl3(bft.Name(), "BFTCoSi handle Challenge PREPARE")
+	// go to response if leaf
+	if bft.IsLeaf() {
+		return bft.startResponsePrepare()
+	}
+	var err error
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, ch)
```

Comment:
>Another question is whether the SendTo should go into a go-routine. Useful if you have lots of children
Do you mean One go routine for sending or One go routine / destinations = child.
I guess the first solution is not necessary because the caller can do it itself very easily if it wants to : 
`go bft.SendToChildren(msg)`
The second solution might be a good candidate, but I would also let the user choose by providing him both methods. One that sends sequentially in ONE go routine and the other by using one goroutine for each destination. What do you think ?